### PR TITLE
Fix: Ensure there is liquidity to release borrowed tokens

### PR DESCRIPTION
**Pull request by ivansolomonoff**
## Security Checklist

- [ ] 1. Re-Entrancy
- [ ] 2. Arithmetic Over/Under Flows
- [ ] 3. Unexpected Ether
- [ ] 4. Delegatecall
- [ ] 5. Default Visibilities
- [ ] 6. Entropy Illusion
- [ ] 7. External Contract Referencing
- [ ] 8. Short Address/Parameter Attack (off chain)
- [ ] 9. Unchecked CALL Return Values
- [ ] 10. Race Conditions / Front Running
- [ ] 11. Denial Of Service (DOS)
- [ ] 12. Block Timestamp Manipulation
- [ ] 13. Constructors with Care
- [ ] 14. Uninitialized Storage Pointers
- [ ] 15. Floating Points and Precision
- [ ] 16. Tx.Origin Authentication
- [ ] 17. Address.isContract Re-Entrancy via Constructor

### ⚠️ NOTES ⚠️

Make sure to think about each of these exploits in this PR.

### CODE REVIEW

File: contracts/protocol/libraries/logic/ValidationLogic.sol [hunk 1]
```diff
@@ -252,6 +252,17 @@ library ValidationLogic {
             Errors.PRICE_ORACLE_SENTINEL_CHECK_FAILED
         );
 
+        if (params.releaseUnderlying) {
+            vars.availableLiquidity = IERC20(params.asset).balanceOf(
+                params.reserveCache.xTokenAddress
```
Comment (GopherJ) [inline contracts/protocol/libraries/logic/ValidationLogic.sol:257]: should take `reserves` into account

File: contracts/protocol/libraries/logic/ValidationLogic.sol [hunk 2]
```diff
@@ -252,6 +252,17 @@ library ValidationLogic {
...
+            );
+
+            require(
```
Comment (yrong) [inline contracts/protocol/libraries/logic/ValidationLogic.sol:260]: if we already check collateral can cover debt [here](https://github.com/para-space/paraspace-core/blob/da9d57291c01f33de106819347a624f4ae84b181/contracts/protocol/libraries/logic/ValidationLogic.sol#L327-L331) do we still need this?

File: contracts/protocol/libraries/logic/ValidationLogic.sol [hunk 3]
```diff
@@ -252,6 +252,17 @@ library ValidationLogic {
...
+                vars.availableLiquidity >= params.amount,
+                Errors.NOT_ENOUGH_LIQUIDITY_IN_RESERVE
+            );
```
Comment (yrong) [inline contracts/protocol/libraries/logic/ValidationLogic.sol:263]: if we already check collateral can always cover debt [here](https://github.com/para-space/paraspace-core/blob/da9d57291c01f33de106819347a624f4ae84b181/contracts/protocol/libraries/logic/ValidationLogic.sol#L327-L331) do we still need this?

File: contracts/protocol/libraries/logic/ValidationLogic.sol (hunk 1 above)
Comment (zhoujia6139) [reply contracts/protocol/libraries/logic/ValidationLogic.sol:257]: I think there is no need to.

File: contracts/protocol/libraries/logic/ValidationLogic.sol (hunk 3 above)
Comment (zhoujia6139) [reply contracts/protocol/libraries/logic/ValidationLogic.sol:263]: Yes, we need. what we are checking here is if cash is enough for borrowing, not LTV.

Comment (ivansolomonoff) [reply contracts/protocol/libraries/logic/ValidationLogic.sol:263]: yeah, basically we want to make sure that when the pool is releasing tokens to the users, that it actually has them.

**Comment by WalidOfNow** [comment]
I personally don't think this is necessary as it adds gas cost to each tx. it can be helpful at the initial stage but then it just becomes an unneeded gas cost.

**Comment by GopherJ** [comment]
@ivansolomonoff talked with the team and we agree with walid that this may not be necessary

**Pull request closed without merging**